Wrappers/Python/cil/optimisation/functions/SubsetSumFunction.py

```diff
@@ -303,39 +297,366 @@ def memory_reset(self):
+        Mathematical Programming (162) pp.83-112. 2017.
+        """
+
+        if not self.gradients_allocated:
+            self.memory_init(x) 
+
+        # Select the next subset (uniformly at random by default). This generates self.subset_num
+        self.next_subset()
+
+        # Compute gradient for current subset and current iterate. store in tmp1
+        # tmp1 = gradient F_{subset_num} (x)
+        self.functions[self.subset_num].gradient(x, out=self.tmp1)
+
+        # Compute the difference between the gradient of function subset_num at current iterate and the subset gradient in memory. store in tmp2
+        # tmp2 = gradient F_{subset_num} (x) - subset_gradient_in_memory_{subset_num}
+        self.tmp1.axpby(1., -1., self.subset_gradients[self.subset_num], out=self.tmp2)
+
+        # Compute the output : 1/num_subsets * tmp2 + full_gradient
+        if out is None:
+            ret = 0.0 * self.tmp2
+            self.tmp2.axpby(1./self.num_subsets, 1., self.full_gradient, out=ret)
+        else:
+            self.tmp2.axpby(1./self.num_subsets, 1., self.full_gradient, out=out)
+
+        # Apply preconditioning
+        if self.precond is not None:
+            if out is None:
+                ret.multiply(self.precond(self.subset_num,x),out=ret)
+            else:
+                out.multiply(self.precond(self.subset_num,x),out=out)
+
+        # Update subset gradients in memory: store the computed gradient F_{subset_num} (x) in self.subset_gradients[self.subset_num]
+        self.subset_gradients[self.subset_num].fill(self.tmp1)
+
+        # Update the full gradient estimator: add 1/num_subsets * (gradient F_{subset_num} (x) - subset_gradient_in_memory_{subset_num}) to the current full_gradient
+        self.full_gradient.axpby(1., 1./self.num_subsets, self.tmp2, out=self.full_gradient)
+
+        if out is None:
+            return ret
+
+class SVRGFunction(SubsetSumFunction):
+    '''Class for use as objective function in gradient type algorithms to enable the use of subsets.
+
+    The `gradient` method does not correspond to the mathematical gradient of a sum of functions, 
+    but rather to a variance-reduced approximated gradient corresponding to the minibatch SVRG algorithm.
+    More details can be found below, in the gradient method.
+
+    Parameters:
+    -----------
+
+    - (optional) 
+        precond: function taking into input an integer (subset_num) and a DataContainer and outputting a DataContainer
+        serves as diagonal preconditioner
+        default None
+    - (optional)
+        gradient_initialisation_point: point to initialize the gradient of each subset
+        and the full gradient
+        default None
+           
+    '''
+
+    def __init__(self, functions, precond=None, update_frequency = 2, **kwargs):
+
+        super(SVRGFunction, self).__init__(functions)
 
         self.gradients_allocated = False
+    
         self.precond = precond
+        self.update_frequency = update_frequency
```

Comment:
Add documentation for `update_frequency`? It is in terms of epochs. Also add a check `update_frequency>0`